test/karma-testharness-adapter.js

```diff
@@ -0,0 +1,186 @@
+      message += 'Stale failure expectations, test no longer exists:\n';
+      message += staleFailureExpectations.map(function(name) { return '  ' + name + '\n'; }).join('');
+      message += '\n';
+    }
+    if (passedUnexpectedly) {
+      message += 'Passed unexpectedly:\n';
+      message += unexpectedPasses.map(function(name) { return '  ' + name + '\n'; }).join('');
+      message += '\n';
+    }
+    if (failedDifferently) {
+      message += 'Failed differently:\n';
+      for (var name in differentFailures) {
+        message += '  Test: ' + stringify(name) + '\n';
+        message += '  Expected: ' + stringify(expectedFailures[name]) + '\n';
+        message += '  Actual:   ' + stringify(differentFailures[name]) + '\n\n';
+      }
+      message += '\n';
+    }
+    if (failedUnexpectedly) {
+      message += 'Failed unexpectedly:\n';
+      for (var name in unexpectedFailures) {
+        message += '  Test: ' + stringify(name) + '\n';
+        message += '  Failure: ' + stringify(unexpectedFailures[name]) + '\n\n';
+      }
+      message += '\n';
+    }
+
+    return {
+      suite: [testURL],
+      description: '',
+      success: message == '',
+      log: message ? [message] : [],
+    };
+  }
+
+  function checkConfig(config) {
+    var requiredProperties = {
+      testURLList: '<Array of test URLs>',
+      skip: '<Object mapping skipped test URLs to the reason for skipping>',
+      expectedFailures: '<Object mapping test URLs to expected inner test failures>',
+      flakyFailureIndicator: '<String used in expectedFailures to indicate non deterministic failure messages>',
+    };
+    var errorMessage = '';
+    if (!config) {
+      errorMessage = 'Missing testharnessTests config options on Karma\'s config.client.\n';
+    } else {
+      for (var property in requiredProperties) {
+        if (!(property in config)) {
+          errorMessage += 'Missing property ' + property + ' from testharnessTests.\n';
+        }
+      }
+    }
+    if (errorMessage) {
+      karma.error('testharnessTests config options must be defined in Karma\'s config.client with the following properties:\n' +
+          Object.keys(requiredProperties).map(function(property) {
+            return '  ' + property + ': ' + requiredProperties[property] + '\n';
+          }).join('') +
+          '\n' +
+          errorMessage);
+      return false;
+    }
+    return true;
+  }
+
+  karma.start = function() {
+    // Karma's config.client object appears as karma.config here.
+    var config = karma.config.testharnessTests;
+    if (!checkConfig(config)) {
+      return;
+    }
+
+    karma.info({total: config.testURLList.length});
+
+    var testName = document.createElement('div');
+    document.body.appendChild(testName);
+    var iframe = document.createElement('iframe');
+    iframe.style.width = 'calc(100vw - 30px)';
+    iframe.style.height = 'calc(100vh - 60px)';
+    document.body.appendChild(iframe);
+
+    function runRemainingTests() {
```

Comment:
I don't understand the control flow for this function. Why is it called recursively? And can the function be defined outside of the definition of karma.start so that it's easier to read / easier to see what's inside runRemainingTests and what's not?